packages/worker/native/src/lib.rs

```diff
@@ -81,118 +141,287 @@ fn worker_new(mut cx: FunctionContext) -> JsResult<BoxedWorker> {
+                                            send_error(queue, callback, err);
+                                        }
+                                    },
+                                    Request::SendActivityHeartbeat {
+                                        heartbeat,
+                                        callback,
+                                    } => match core.send_activity_heartbeat(heartbeat).await {
+                                        Ok(()) => {
+                                            send_result(queue, callback, |cx| Ok(cx.undefined()));
+                                        }
+                                        Err(err) => {
+                                            send_error(queue, callback, err);
+                                        }
+                                    },
+                                    // Ignore BreakPoller and Shutdown, they're handled above
+                                    Request::BreakPoller => {}
+                                    Request::Shutdown => {}
+                                }
+                            });
+                        }
+                    }
+                    Err(err) => {
+                        send_error(queue.clone(), callback, err);
+                    }
                 }
-            }
-        }
+            })
     });
 
-    Ok(cx.boxed(cloned_worker))
+    Ok(cx.undefined())
 }
 
-/// Initiate a single poll request.
-/// Will block if a poll request is already in-flight
-fn worker_poll(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+/// Initiate a single workflow activation poll request.
+/// There should be only one concurrent poll request for this type.
+fn worker_poll_workflow_activation(mut cx: FunctionContext) -> JsResult<JsUndefined> {
     let worker = cx.argument::<BoxedWorker>(0)?;
     let queue_name = cx.argument::<JsString>(1)?.value(&mut cx);
-    let callback = cx.argument::<JsFunction>(2)?.root(&mut cx);
-    let item = PollRequest {
+    let callback = cx.argument::<JsFunction>(2)?;
+    let request = Request::PollWorkflowActivation {
         queue_name,
-        callback,
+        callback: callback.root(&mut cx),
     };
-    match worker.sender.send(item) {
-        Ok(_) => Ok(cx.undefined()),
-        Err(err) => {
-            let error = JsError::error(&mut cx, format!("{}", err))?;
-            cx.throw(error)
-        }
+    if let Err(err) = worker.sender.blocking_send(request) {
+        callback_with_error(&mut cx, callback, err)?;
+    }
+    Ok(cx.undefined())
+}
+
+/// Initiate a single activity task poll request.
+/// There should be only one concurrent poll request for this type.
+fn worker_poll_activity_task(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+    let worker = cx.argument::<BoxedWorker>(0)?;
+    let queue_name = cx.argument::<JsString>(1)?.value(&mut cx);
+    let callback = cx.argument::<JsFunction>(2)?;
```

Comment:
You could impl some kind of `unpack` function on `FunctionContext` that would de-dupe these preambles and make things a bit more readable

Reply:
@bergundy Something like this
```rust
trait FnContextExt {
  fn unpack(self) -> (BoxedWorkerType, QueueNameType, CallbackType);
}

impl FnContextExt for FunctionContext {
    /// Implement the three lines here and return 3-tuple (or a real struct)
}
```